src/System.Net.Http/src/System/Net/Http/Headers/HeaderUtilities.cs

```diff
@@ -249,14 +249,69 @@ internal static void CheckValidQuotedString(string value, string parameterName)
             return null;
         }
 
-        internal static bool TryParseInt32(string value, out int result)
+        internal static bool TryParseInt32(string value, out int result) =>
+            TryParseInt32(value, 0, value.Length, out result);
+
+        internal static bool TryParseInt32(string value, int offset, int length, out int result)
         {
-            return int.TryParse(value, NumberStyles.None, NumberFormatInfo.InvariantInfo, out result);
+            // TODO #21281: Replace with int.TryParse(Span<char>) once it's available
+
+            if (offset < 0 || length < 0 || offset > value.Length - length)
+            {
+                result = 0;
+                return false;
+            }
+
+            int tmpResult = 0;
+            int pos = offset;
+            for (int i = 0; i < length; i++)
+            {
+                char c = value[pos++];
+                int digit = c - '0';
+                if ((uint)digit > 9 || // invalid digit
+                    tmpResult > int.MaxValue / 10 || // will overflow once digit is added in
+                    (tmpResult == int.MaxValue / 10 && digit > 7)) // will overflow once digit is added in
+                {
+                    // Invalid digit or about to overflow
+                    result = 0;
+                    return false;
+                }
+                tmpResult = (tmpResult * 10) + digit;
```

Comment:
> possibly second check is not necessary

It is.  Consider tmpResult=1288490189... tmpResult*10 overflows multiple times and results in 2, so the first check wouldn't detect the overflow.

> you can check for overflow in more readable way like this

That's slower, e.g. this version has a division per digit at run-time (the division in the PR is done at compile time).